File: app/src/main/java/com/sjk/deleterecentpictures/MainActivity.java
package com.sjk.deleterecentpictures;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.PendingIntent;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.drawable.Icon;
import android.media.ThumbnailUtils;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.CancellationSignal;
import android.os.Environment;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.util.Size;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.pm.ShortcutInfoCompat;
import androidx.core.content.pm.ShortcutManagerCompat;
import androidx.preference.PreferenceManager;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.sjk.deleterecentpictures.utils.FileUtil;
import com.sjk.deleterecentpictures.utils.ImageScanner;
import com.sjk.deleterecentpictures.utils.ShortcutUtil;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    
    public static List<String> imagePaths;
    //    private static String imagePath;
    public static List<Bitmap> theLatestImages;
//    public static Bitmap theLatestImage;
    
    private final static String[] PERMISSIONS_STORAGE = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
    };
    
    @SuppressLint("HandlerLeak")
    protected final Handler handler = new Handler() {
        @Override
        public void handleMessage(@NonNull Message msg) {
            switch (Objects.requireNonNull(MainActivityHandlerMsgWhat.getByValue(msg.what))) {
                case ERROR: {
                    //错
                    String string = msg.obj.toString();
                    Toast.makeText(getApplicationContext(), "出错了：" + string, Toast.LENGTH_SHORT).show();
                    break;
                }
                case REFRESH_TEXT: {
                    //设置文本
//                    String imagePath = msg.obj.toString();
//                    MainActivity.imagePath = imagePath;
                    Button latestPicturePathButton = findViewById(R.id.latestPicturePathButton);
                    latestPicturePathButton.setText(imagePaths.get(0));
                    break;
                }
                case REFRESH_IMAGE: {
                    //设置图片
                    Bitmap bitmap = (Bitmap) msg.obj;
                    ImageView latestPictureImageView = findViewById(R.id.latestPictureImageView);
                    latestPictureImageView.setImageBitmap(bitmap);
                    latestPictureImageView.setContentDescription(imagePaths.get(0));
                    break;
                }
                case DELETE_IMAGE_SUCCESS: {
                    Toast.makeText(getApplicationContext(), "删除图片 " + imagePaths.get(0) + " 成功", Toast.LENGTH_SHORT).show();
                    SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(MainActivity.this);
                    boolean closeApp = sp.getBoolean("closeApp", true);
                    if (closeApp) {
                        finish();
                    } else {
                        read();
                    }
                    break;
                }
                case DELETE_IMAGE_FAIL: {
                    Toast.makeText(getApplicationContext(), "出错了：删除图片 " + imagePaths.get(0) + " 失败", Toast.LENGTH_SHORT).show();
                    break;
                }
                default: {
                    break;
                }
            }
        }
    };
    
    @SuppressLint("WrongConstant")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        
        //设置默认偏好
        PreferenceManager.setDefaultValues(this, R.xml.root_preferences, false);
        
        initView();
        requestWritePermission();
    }
    
    private void initView() {
        setContentView(R.layout.activity_main);
        buttonClickEventBind();
    }
    
    private void initList() {
        theLatestImages = new ArrayList<>();
        imagePaths = new ArrayList<>();
    }
    
    private void buttonClickEventBind() {
        Button latestPicturePathButton = findViewById(R.id.latestPicturePathButton);
        latestPicturePathButton.setOnClickListener(v -> {
            ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
            ClipData clip = ClipData.newPlainText(getResources().getString(R.string.app_name), imagePaths.get(0));
            if (clipboard != null) {
                clipboard.setPrimaryClip(clip);
                Toast.makeText(getApplicationContext(), "已复制到剪切板", Toast.LENGTH_SHORT).show();
            } else {
                Message message = new Message();
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "复制失败";
                handler.sendMessage(message);
            }
//            createShortcut();
        });
        latestPicturePathButton.setOnLongClickListener(v -> {
            Toast.makeText(this, "当前查找目录下有" + imagePaths.size() + "张图片", Toast.LENGTH_SHORT).show();
            return true;
        });
        
        Button refreshButton = findViewById(R.id.refreshButton);
        refreshButton.setOnClickListener(v -> {
            Toast.makeText(this, "刷新", Toast.LENGTH_SHORT).show();
            read();
        });
        
        Button openImageActivityButton = findViewById(R.id.openImageActivityButton);
        openImageActivityButton.setOnClickListener(v -> {
            //打开图片查看界面
            Intent intent = new Intent(getApplicationContext(), ImageActivity.class);
            startActivity(intent);
        });
        
        Button cancelButton = findViewById(R.id.cancelButton);
        cancelButton.setOnClickListener(v -> finish());
        
        //删除按钮
        Button deleteButton = findViewById(R.id.deleteButton);
        deleteButton.setOnClickListener(v -> {
            SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(MainActivity.this);
            boolean deleteDirectly = sp.getBoolean("deleteDirectly", false);
            if (deleteDirectly) {
                onDeleteButtonClick();
            } else {
                new MaterialAlertDialogBuilder(MainActivity.this)
                        .setTitle("警告")
                        .setMessage("请确认是否删除\n" + imagePaths.get(0))
                        .setPositiveButton("确定", (dialog, which) -> onDeleteButtonClick())
                        .setNegativeButton("取消", (dialog, which) -> dialog.cancel())
                        .show();
            }
        });
        deleteButton.setOnLongClickListener(v -> {
            onDeleteButtonClick();
            finish();
            return true;
        });
        
        Button settingsButton = findViewById(R.id.settingsButton);
        settingsButton.setOnClickListener(v -> {
            Intent intent = new Intent(MainActivity.this, SettingsActivity2.class);
            startActivityForResult(intent, 1);
        });
    }
    
    /**
     * 点击删除按钮触发事件
     */
    protected void onDeleteButtonClick() {
        new Thread(() -> {
//          Log.d("imagePath", "run: " + imagePaths.get(0));
            if (imagePaths.get(0) != null && !imagePaths.get(0).equals("")) {
                Message message = new Message();
                //删除图片并判断
                if (FileUtil.deleteFile(imagePaths.get(0))) {
                    message.what = MainActivityHandlerMsgWhat.DELETE_IMAGE_SUCCESS.getIndex();
                    FileUtil.updateFileFromDatabase(getApplicationContext(), imagePaths.get(0));
                } else {
                    message.what = MainActivityHandlerMsgWhat.DELETE_IMAGE_FAIL.getIndex();
                }
                handler.sendMessage(message);
            } else {
                Message message = new Message();
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "没有获取到图片路径，删除失败";
                handler.sendMessage(message);
            }
        }).run();
    }
    
    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        Log.d(TAG, "onActivityResult: " + requestCode + " " + requestCode + " " + data);
        if (requestCode == 1 && resultCode == RESULT_OK && data != null) {
            boolean result = data.getBooleanExtra("preferenceChanged", false);
            if (result) {
                startActivity(new Intent(MainActivity.this, MainActivity.class));
                finish();
                
                Toast.makeText(this, "已重新加载", Toast.LENGTH_SHORT).show();
            }
        } else {
//            Toast.makeText(this, "无返回值", Toast.LENGTH_SHORT).show();
        }
        
        super.onActivityResult(requestCode, resultCode, data);
    }
    
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == 0) {
            if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
                Message message = new Message();
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "没有获取到存储权限，自动退出";
                handler.sendMessage(message);
                finish();
                return;
            }
        }
//        initView();
        read();
    }
    
    private void requestWritePermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, 0);
            } else {
//                initView();
                read();
            }
        } else {
            read();
        }
    }
    
    protected void read() {
        new Thread(() -> {
            initList();
            Message message = new Message();
            imagePaths = ImageScanner.getImages(getApplicationContext(), getSelection(), true);
            if (imagePaths == null) {
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "查询路径不合法，请重新设置";
                handler.sendMessage(message);
                return;
            }
            
            if (imagePaths.size() == 0) {
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "没有找到图片";
                handler.sendMessage(message);
                return;
            }
            
            String imagePath = imagePaths.get(0);
            if (imagePath != null) {
                Message textViewMessage = new Message();
                textViewMessage.what = MainActivityHandlerMsgWhat.REFRESH_TEXT.getIndex();
//                        textViewMessage.obj = imagePath;
                handler.sendMessage(textViewMessage);
                
                Bitmap bitmap;
                try {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);
                        int thumbnailSize = sp.getInt("thumbnailSize", 512);
                        bitmap = ThumbnailUtils.createImageThumbnail(
                                new File(imagePath),
                                new Size(thumbnailSize, thumbnailSize),
                                new CancellationSignal()
                        );
                    } else {
                        bitmap = BitmapFactory.decodeFile(imagePaths.get(0));
                    }
                } catch (Exception e) {
                    e.printStackTrace();
                    message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                    message.obj = "读取图片失败";
                    handler.sendMessage(message);
                    return;
                }
                
                message.what = MainActivityHandlerMsgWhat.REFRESH_IMAGE.getIndex();
                message.obj = bitmap;
                handler.sendMessage(message);
                
            } else {
                message.what = MainActivityHandlerMsgWhat.ERROR.getIndex();
                message.obj = "查找图片失败";
                handler.sendMessage(message);
            }
        }
        ).run();
    }
    
    private String getSelection() {
        String[] strings = getResources().getStringArray(R.array.path_values);
        SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(this);
        String pathType = sp.getString("path", strings[0]);
//        Log.d(TAG, "read: " + sp.getString("path", strings[0]));
        
        String selection = null;
        if (strings[0].equals(pathType)) {
            selection = null;
        } else if (strings[1].equals(pathType)) {
            selection = ImageScanner.getScreenshotsPath();
        } else if (strings[2].equals(pathType)) {
            selection = Environment.getExternalStorageDirectory().getPath() + "/" + sp.getString("customizePath", "");
        }
        Log.d(TAG, "getSelection: 查询目录" + selection);
        return selection;
    }
    
    /*public static Bitmap bitmapCompress(Bitmap image) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        image.compress(Bitmap.CompressFormat.JPEG, 100, baos);
        int options = 100;
        while (baos.toByteArray().length / 1024 > 10) {
            baos.reset();
            image.compress(Bitmap.CompressFormat.JPEG, options, baos);
            options -= 10;
            if (options <= 0) {
                break;
            }
        }
        ByteArrayInputStream isBm = new ByteArrayInputStream(baos.toByteArray());
        return BitmapFactory.decodeStream(isBm, null, null);
    }*/
    
    enum MainActivityHandlerMsgWhat {
        //错误
        ERROR(-1),
        //刷新文本
        REFRESH_TEXT(0),
        //刷新图片
        REFRESH_IMAGE(1),
        //删除图片成功
        DELETE_IMAGE_SUCCESS(2),
        //删除图片失败
        DELETE_IMAGE_FAIL(3);
        
        private final int index;
        
        MainActivityHandlerMsgWhat(int index) {
            this.index = index;
        }
        
        public int getIndex() {
            return index;
        }
        
        public static MainActivityHandlerMsgWhat getByValue(int what) {
            for (MainActivityHandlerMsgWhat handlerMsgWhat : values()) {
                if (handlerMsgWhat.getIndex() == what) {
                    return handlerMsgWhat;
                }
            }
            return null;
        }
    }
}

File: app/src/main/java/com/sjk/deleterecentpictures/MainActivity.kt
package com.sjk.deleterecentpictures


import android.Manifest
import android.app.Activity
import android.content.*
import android.content.pm.PackageManager
import android.net.Uri
import android.os.*
import android.provider.Settings
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import androidx.core.app.ActivityCompat
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewpager2.widget.ViewPager2
import com.davemorrissey.labs.subscaleview.ImageSource
import com.davemorrissey.labs.subscaleview.SubsamplingScaleImageView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.sjk.deleterecentpictures.common.*
import com.sjk.deleterecentpictures.utils.FileUtil
import com.sjk.deleterecentpictures.utils.ImageScannerUtil
import pl.droidsonroids.gif.GifDrawable
import pl.droidsonroids.gif.GifImageView
import java.util.*


open class MainActivity : BaseActivity() {
    
    private var isLoaded = false
    private lateinit var viewPager: ViewPager2
    private val viewPagerAdapter = MainActivityViewPagerAdapter()
    private var currentImagePath: String? = null
    private val event: Event = App.newEvent
    private var viewPagerCurrentPosition = 0
    
    companion object {
        private const val TAG = "MainActivity"
        
        //    public static Bitmap theLatestImage;
        private val PERMISSIONS_STORAGE = arrayOf(
                Manifest.permission.READ_EXTERNAL_STORAGE,
                Manifest.permission.WRITE_EXTERNAL_STORAGE,
                Manifest.permission.MANAGE_EXTERNAL_STORAGE,
        )
    }
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        
        //设置默认偏好
        PreferenceManager.setDefaultValues(this, R.xml.root_preferences, false)
        
        initView()
        requestWritePermission()
    }
    
    private fun initView() {
        setContentView(R.layout.activity_main)
        buttonClickEventBind()
        
        this.viewPager = findViewById(R.id.viewPager)
        this.viewPager.adapter = viewPagerAdapter
    }
    
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (this.isLoaded) {
            return
        }
        
        isLoaded = true
    }
    
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        Log.d(TAG, "onActivityResult: $requestCode $requestCode $data")
        if (requestCode == 1 && resultCode == Activity.RESULT_OK && data != null) {
            val result = data.getBooleanExtra("preferenceChanged", false)
            if (result) {
                startActivity(Intent(this@MainActivity, MainActivity::class.java))
                finish()
                this.getOutPut().showToast("已重新加载设置")
            }
        }/* else {
//            Toast.makeText(this, "无返回值", Toast.LENGTH_SHORT).show();
        }*/
        super.onActivityResult(requestCode, resultCode, data)
    }
    
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == 0) {
            if (grantResults[0] == PackageManager.PERMISSION_DENIED) {
                this.getOutPut().showToast("未获取到存储权限")
                logW(TAG, "Storage permission not obtained")
                finish()
                return
            }
        }
        
        this.refreshAll()
    }
    
    private fun requestWritePermission() {
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (!Environment.isExternalStorageManager()) {
                val intent = Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION)
                intent.data = Uri.parse("package:$packageName")
                startActivityForResult(intent, 1)
            }
        }
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PERMISSIONS_STORAGE.iterator().forEach {
                if (checkSelfPermission(it) != PackageManager.PERMISSION_GRANTED) {
                    ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, 0)
                    return
                }
            }
            this.refreshAll()
            return
        }
        
        this.refreshAll()
    }
    
    private fun buttonClickEventBind() {
        val latestPicturePathButton = findViewById<Button>(R.id.latestPicturePathButton)
        latestPicturePathButton.setOnClickListener {
            if (this.currentImagePath == null) {
                return@setOnClickListener
            }
            this.getOutPut().showToastLong("完整路径：${this.currentImagePath}")
        }
        latestPicturePathButton.setOnLongClickListener {
            if (this.currentImagePath == null) {
                this.getOutPut().showToast("无路径")
                return@setOnLongClickListener true
            }
            
            val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
            val clip = ClipData.newPlainText(resources.getString(R.string.app_name), this.currentImagePath)
            clipboard.setPrimaryClip(clip)
            this.getOutPut().showToast("${this.currentImagePath}已复制到剪切板")
            true
        }
        val refreshButton = findViewById<Button>(R.id.refreshButton)
        refreshButton.setOnClickListener {
            this.refreshAll()
            this.getOutPut().showToast("刷新成功")
        }
        /*val openImageActivityButton = findViewById<Button>(R.id.openImageActivityButton)
        openImageActivityButton.setOnClickListener {
            //打开图片查看界面
            val intent = Intent(applicationContext, ImageActivity::class.java)
            startActivity(intent)
            overridePendingTransition(android.R.anim.fade_in, android.R.anim.fade_out);
        }*/
        val cancelButton = findViewById<Button>(R.id.cancelButton)
        cancelButton.setOnClickListener { finish() }
        
        //删除按钮
        val deleteButton = findViewById<Button>(R.id.deleteButton)
        deleteButton.setOnClickListener {
            
            val deleteDirectly = this.getDataSource().getSP().getBoolean("deleteDirectly", false)
            if (deleteDirectly) {
                deleteCurrentImage()
            } else {
                MaterialAlertDialogBuilder(this@MainActivity)
                        .setTitle("提示")
                        .setMessage("请确认是否删除\n${this.currentImagePath}")
                        .setPositiveButton("确定") { _: DialogInterface?, _: Int -> deleteCurrentImage() }
                        .setNegativeButton("取消") { dialog: DialogInterface, _: Int -> dialog.cancel() }
                        .show()
            }
        }
        deleteButton.setOnLongClickListener {
            deleteCurrentImage(false)
            finish()
            true
        }
        val settingsButton = findViewById<Button>(R.id.settingsButton)
        settingsButton.setOnClickListener {
            val intent = Intent(this@MainActivity, SettingsActivity2::class.java)
            startActivityForResult(intent, 1)
        }
    }
    
    private fun deleteCurrentImage(needToRefresh: Boolean = true) {
        if (this.currentImagePath == null || this.currentImagePath == "") {
            this.getOutPut().showToast("没有获取到图片路径，删除失败")
            return
        }
        
        val deleteButton: Button = findViewById(R.id.deleteButton)
        deleteButton.isEnabled = false
        
        val scanType: String = this.getDataSource().getSP().getString("scanType", "1")!!
        
        if (!App.fileUtil.existsFile(this.currentImagePath)) {
            this.getOutPut().showToast("文件不存在，删除失败")
            ImageScannerUtil.refreshMediaLibraryByPath(applicationContext, this.currentImagePath, scanType.toInt()) { path: String, uri: Uri? ->
                // 里面是线程回调，这里面不能刷新UI，除非放到UI线程里执行
                this.runOnUiThread {
                    this.refreshImages()
                    deleteButton.isEnabled = true
                }
            }
            return
        }
        
        //删除图片并判断
        if (FileUtil.deleteFile(this.currentImagePath)) {
            this.getOutPut().showToast("${this.currentImagePath}删除成功")
        } else {
            this.getOutPut().showToast("文件无法删除，请确认是否已给予存储权限")
        }
        ImageScannerUtil.refreshMediaLibraryByPath(applicationContext, this.currentImagePath, scanType.toInt()) { path: String, uri: Uri? ->
            // 里面是线程回调，这里面不能刷新UI，除非放到UI线程里执行
            this.runOnUiThread {
                if (this.getDataSource().getSP().getBoolean("closeApp", true)) {
                    this.finish()
                    return@runOnUiThread
                }
                if (needToRefresh) {
                    this.refreshImages()
                    this.refreshCurrentImagePath()
                }
                deleteButton.isEnabled = true
            }
        }
    }
    
    private fun refreshAll() {
        Thread {
            this.refreshImages()
            this.viewPager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
                override fun onPageSelected(position: Int) {
                    super.onPageSelected(position)
                    this@MainActivity.viewPagerCurrentPosition = position
        
        
                    if (viewPagerAdapter.imagePaths.size == 0) {
                        this@MainActivity.currentImagePath = null
                        return
                    }
        
                    this@MainActivity.refreshCurrentImagePath()
                }
            })
        }.start()
    }
    
    private fun refreshCurrentImagePath() {
        this@MainActivity.currentImagePath = this.viewPagerAdapter.imagePaths[this.viewPagerCurrentPosition]
        val latestPicturePathButton = findViewById<Button>(R.id.latestPicturePathButton)
        latestPicturePathButton.text = App.dataSource.getSimplifiedPathInExternalStorage(this@MainActivity.currentImagePath)
    }
    
    private fun refreshImages() {
        ImageScannerUtil.init(this, this.getDataSource().getSelection())
        this.viewPagerAdapter.imagePaths = arrayListOf()
        
        var i = this.getDataSource().getNumberOfPictures()
        while (i > 0) {
            val imagePath = App.imageScannerUtil.getNext() ?: break
            if (!App.fileUtil.existsFile(imagePath)) {
                continue
            }
            this.viewPagerAdapter.imagePaths.add(imagePath)
            i--
        }
        /*for (i in 0..this.getDataSource().getNumberOfPictures()) {
            val imagePath = App.imageScannerUtil.getNext() ?: break
            this.viewPagerAdapter.imagePaths.add(imagePath)
        }*/
        
        if (this.viewPagerAdapter.imagePaths.size == 0) {
//            this.getOutPut().showToast("未发现图片")
            val latestPicturePathButton = findViewById<Button>(R.id.latestPicturePathButton)
            latestPicturePathButton.text = "未发现图片"
        }
//        this.viewPager.adapter = viewPagerAdapter
        this.runOnUiThread {
            this.viewPagerAdapter.notifyDataSetChanged()
            this.refreshCurrentImagePath()
        }
    }
}

internal class MainActivityViewPagerAdapter : RecyclerView.Adapter<MainActivityViewPagerAdapter.ViewPagerViewHolder>() {
    private var viewPagerViewHolders: MutableList<ViewPagerViewHolder> = ArrayList()
    var imagePaths: MutableList<String?> = arrayListOf()
    val event: Event = App.newEvent
    
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewPagerViewHolder {
        val viewPagerViewHolder = ViewPagerViewHolder(LayoutInflater.from(parent.context)
                .inflate(R.layout.layout_main_view_pager_item, parent, false))
        viewPagerViewHolders.add(viewPagerViewHolder)
        return viewPagerViewHolder
    }
    
    fun add(index: Int, path: String?) {
        imagePaths.add(index, path)
//        this.notifyDataSetChanged()
    }
    
    fun addLast(path: String?) {
        imagePaths.add(path)
//        this.notifyDataSetChanged()
    }
    
    fun addFirst(path: String?) {
        imagePaths.add(0, path)
//        this.notifyDataSetChanged()
    }
    
    fun remove(index: Int) {
        imagePaths.removeAt(index)
//        this.notifyDataSetChanged()
    }
    
    fun removeLast() {
        imagePaths.removeAt(imagePaths.size - 1)
//        this.notifyDataSetChanged()
    }
    
    fun removeFirst() {
        imagePaths.removeAt(0)
//        this.notifyDataSetChanged()
    }
    
    override fun onBindViewHolder(holder: ViewPagerViewHolder, position: Int) {
        holder.imagePath = imagePaths[position]
        if (holder.imagePath == null) {
            holder.latestPictureGifImageView.visibility = View.GONE
            holder.gifSign.visibility = View.GONE
            holder.latestPictureImageView.visibility = View.GONE
            return
        }
        
        if (!App.fileUtil.existsFile(holder.imagePath)) {
            return
        }
        
        if (position < imagePaths.size) {
            holder.isGif = FileUtil.isGifFile(holder.imagePath)
            if (holder.isGif) {
                holder.latestPictureGifImageView.visibility = View.VISIBLE
                holder.gifSign.visibility = View.VISIBLE
                holder.latestPictureImageView.visibility = View.GONE
                
                return
            }
            holder.latestPictureGifImageView.visibility = View.GONE
            holder.gifSign.visibility = View.GONE
            holder.latestPictureImageView.visibility = View.VISIBLE
        }
    }
    
    override fun onViewDetachedFromWindow(holder: ViewPagerViewHolder) {
        super.onViewDetachedFromWindow(holder)
        
        if (!holder.isGif) {
            holder.latestPictureImageView.recycle()
            return
        }
        
        val gifDrawable = holder.latestPictureGifImageView.drawable as GifDrawable
        if (gifDrawable.isRecycled) {
            return
        }
        
        gifDrawable.recycle()
    }
    
    override fun onViewAttachedToWindow(holder: ViewPagerViewHolder) {
        super.onViewAttachedToWindow(holder)

//        if (holder.hasImage){
//            return
//        }
        if (holder.imagePath == null) {
            App.outPut.showToast("文件路径为空")
            return
        }
        
        if (!App.fileUtil.existsFile(holder.imagePath)) {
            App.outPut.showToast("文件不存在，可能已被其他软件删除")
            return
        }
        
        if (!holder.isGif) {
            holder.latestPictureImageView.setImage(ImageSource.uri(holder.imagePath!!))
            return
        }
        
        val gifFromPath = GifDrawable(holder.imagePath!!)
        holder.latestPictureGifImageView.setImageDrawable(gifFromPath)
        gifFromPath.start()
    }
    
    override fun getItemCount(): Int {
        return this.imagePaths.size
    }
    
    internal inner class ViewPagerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var latestPictureImageView: SubsamplingScaleImageView = itemView.findViewById(R.id.latestPictureImageView)
        var latestPictureGifImageView: GifImageView = itemView.findViewById(R.id.latestPictureGifImageView)
        var gifSign: View = itemView.findViewById(R.id.gifSign)
        var imagePath: String? = null
        var isGif = false
        
        init {
            val openImageActivityButton = itemView.findViewById<Button>(R.id.openImageActivityButton)
            openImageActivityButton.setOnClickListener {
                //打开图片查看界面
                if (!App.fileUtil.existsFile(this.imagePath)) {
                    App.outPut.showToast("图片无法查看")
                    return@setOnClickListener
                }
                
                App.globalData.setData("currentImagePath", this.imagePath)
                val intent = Intent(itemView.context, ImageActivity::class.java)
                itemView.context.startActivity(intent)
            }
        }
    }
}